File: 백준/Gold/1167. 트리의 지름/트리의 지름.java
import java.io.*;
import java.util.*;

public class Main {

    static int V;
    static int[] head, to, weight, next;
    static int idx = 0;
    static int maxLen = 0;
    static int[] visited;


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        V = Integer.parseInt(br.readLine());

        // 초기화
        head = new int[V + 1];
        Arrays.fill(head, -1);
        to = new int[2 * V];
        weight = new int[2 * V];
        next = new int[2 * V];
        visited = new int[V + 1];

        for (int i = 1; i <= V; i++) {
            String[] parts = br.readLine().split(" ");
            int node = Integer.parseInt(parts[0]);
            for (int j = 1; j < parts.length - 1; j += 2) {
                int v = Integer.parseInt(parts[j]);
                int w = Integer.parseInt(parts[j + 1]);
                addEdge(node, v, w);
            }
        }

        dfs(1);
        System.out.println(maxLen);
    }

    static void addEdge(int u, int v, int w) {
        to[idx] = v;
        weight[idx] = w;
        next[idx] = head[u];
        head[u] = idx++;
    }

    static int dfs(int node) {
        int firstMax = 0, secondMax = 0;
        visited[node] = 1;

        for (int e = head[node]; e != -1; e = next[e]) {
            int v = to[e], w = weight[e];
            if (visited[v] == 0) {
                int pathLen = dfs(v) + w;
                if (pathLen > firstMax) {
                    secondMax = firstMax;
                    firstMax = pathLen;
                } else if (pathLen > secondMax) {
                    secondMax = pathLen;
                }
            }
        }

        maxLen = Math.max(maxLen, firstMax + secondMax);
        return firstMax;
    }
}

File: 백준/Gold/1167. 트리의 지름/트리의 지름.py
import sys
sys.setrecursionlimit(10**6)

V = int(sys.stdin.readline())
adj_lst = [[] for _ in range(V + 1)]
visited = [False] * (V + 1)
max_len = 0

for _ in range(V):
    parts = list(map(int, sys.stdin.readline().split()))
    node = parts[0]
    idx = 1
    while parts[idx] != -1:
        to = parts[idx]
        length = parts[idx + 1]
        adj_lst[node].append((to, length))
        idx += 2

def dfs(node):
    global max_len
    visited[node] = True
    first_max, second_max = 0, 0
    for to, length in adj_lst[node]:
        if not visited[to]:
            path_len = dfs(to) + length
            if path_len > first_max:
                second_max = first_max
                first_max = path_len
            elif path_len > second_max:
                second_max = path_len
    max_len = max(max_len, first_max + second_max)
    return first_max

dfs(1)
print(max_len)